weave-js/src/components/Button/Button.tsx

```diff
@@ -4,153 +4,153 @@
+  ghost:
+    'bg-oblivion/[0.05] dark:bg-moonbeam/[0.05] text-moon-800 dark:text-moon-200 hover:bg-teal-300/[0.48] hover:text-teal-600 dark:hover:bg-teal-700/[0.48] dark:hover:text-teal-400',
+  quiet:
+    'text-moon-500 hover:text-moon-800 dark:hover:text-moon-200 hover:bg-oblivion/[0.05] dark:hover:bg-moonbeam/[0.05]',
+  destructive: 'bg-red-500 text-white hover:bg-red-450',
+};
+
+export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
   (
     {
       size = 'medium',
       variant = 'primary',
       icon,
-      startIcon = icon,
+      startIcon,
       endIcon,
       active,
       children,
       className = '',
       tooltip,
       tooltipProps,
-      twWrapperStyles = {},
+      twWrapperStyles,
       ...htmlAttributes
     },
     ref
   ) => {
-    const hasIcon = startIcon || endIcon;
-    if (!children && !hasIcon) {
-      console.error('Button: requires either children or an icon.');
-    }
-
-    const isSmall = size === 'small';
-    const isMedium = size === 'medium';
-    const isLarge = size === 'large';
-    const isPrimary = variant === 'primary';
-    const isSecondary = variant === 'secondary';
-    const isGhost = variant === 'ghost';
-    const isQuiet = variant === 'quiet';
-    const isDestructive = variant === 'destructive';
-
-    const hasBothIcons = startIcon && endIcon;
-    const hasOnlyOneIcon = hasIcon && !hasBothIcons;
-    const isIconOnly = hasOnlyOneIcon && !children;
-
     /**
      * The Tailwind wrapper is a div, which is a block-element. However, a normal button
      * wouldn't be wrapped and would be inline-block by default. These styles are a
      * necessary workaround to ensure Button is styled correctly despite the wrapper.
      */
-    const wrapperStyles = {
+    const wrapperStyles: CSSProperties = {
       display: 'inline-flex',
       width: className.includes('w-full') ? '100%' : undefined,
       ...twWrapperStyles,
     };
 
-    const button = (
+    /**
+     * Runtime validation of props' type constraints
+     */
+    const ButtonInner = () => {
+      if (icon) {
+        if (children || startIcon || endIcon) {
+          console.warn(
+            'Button: `startIcon`, `endIcon`, and `children` are ignored when using `icon`.'
+          );
+        }
+        return <Icon name={icon} role="presentation" />;
+      }
+
+      if (!children) {
+        console.error('Button: `children` is required when not using `icon`.');
+        return null;
+      }
+
+      return (
+        <>
+          {startIcon ? <Icon name={startIcon} role="presentation" /> : null}
+          {children}
+          {endIcon ? <Icon name={endIcon} role="presentation" /> : null}
+        </>
+      );
+    };
+
+    const ButtonInternal = () => (
       <button
         ref={ref}
+        {...htmlAttributes}
```

Comment:
We want to override the html attrs with our explicitly defined ones so this should come first.

Reply:
This is somewhat of a subjective thing. I'm personally in favor of reducing options and adding them as the need arises rather than starting wide open.